docs/administrators/policy_cookbook.md

```diff
@@ -568,3 +568,226 @@ pep_api_data_obj_close_post(*INSTANCE_NAME, *COMM, *OPENED_DATA_OBJ_INPUT)
+                "data_owner_zone": <string>,
+                "data_is_dirty": <string>,
+                "data_status": <string>,
+                "data_checksum": <string>,
+                "data_expiry_ts": <string>,
+                "data_map_id": <string>,
+                "data_mode": <string>,
+                "r_comment": <string>,
+                "create_ts": <string>,
+                "modify_ts": <string>,
+                "resc_id": <string>
+            },
+            "after": {
+                "data_id": <string>,
+                "coll_id": <string>,
+                "data_repl_num": <string>,
+                "data_version": <string>,
+                "data_type_name": <string>,
+                "data_size": <string>,
+                "data_path": <string>,
+                "data_owner_name": <string>,
+                "data_owner_zone": <string>,
+                "data_is_dirty": <string>,
+                "data_status": <string>,
+                "data_checksum": <string>,
+                "data_expiry_ts": <string>,
+                "data_map_id": <string>,
+                "data_mode": <string>,
+                "r_comment": <string>,
+                "create_ts": <string>,
+                "modify_ts": <string>,
+                "resc_id": <string>
+            },
+            "file_modified": {
+                <string>: <string>,
+                ...
+            }
+        },
+        ...
+    ]
+}
+```
+Each entry in the list of replicas is a "before" and "after" view of the row in `R_DATA_MAIN` representing that replica, complete with the column names. In order to check that the replica has a checksum, we want to find the replica which was modified in the list, and ensure that the "after" entry's "data_checksum" entry has a value. If not, we will calculate a checksum and register it in the catalog.
+
+The `data_object_finalize` database PEP is invoked twice per data transfer: once on open in order to lock the data object, and once on close in order to finalize the system metadata. We are only interested in the invocation on close. You will find in the implementation checks which can detect which is being invoked. You might consider abstracting this into a separate function for more readable code.
+
+Here is the PEP implementation for the example policy:
+```python
+pep_database_data_object_finalize_post(*instance, *ctx, *out, *json_input)
+{
+    # This PEP calculates the checksum of newly created replicas if one does not already exist.
+
+    # Get a JSON handle so we can work with the json_input.
+    msiStrlen(*json_input, *size);
+    msi_json_parse(*json_input, int(*size), *handle);
+
+    # Get the logical path based on the data ID using Language Integrated GenQuery.
+    msi_json_value(*handle, "/replicas/0/before/data_id", *data_id);
+    foreach (*result in select COLL_NAME, DATA_NAME where DATA_ID = '*data_id') {
+        *logical_path = *result.COLL_NAME ++ '/' ++ *result.DATA_NAME;
+        break;
+    }
```

Comment:
Could add some error checking after this in case `*logical_path` is never set.

Reply:
Done